File: 2024/cpp/day-06/main.cpp
#include <iostream>
#include <set>
#include <vector>

using namespace std;

void readInput(vector<string> &mappedArea, pair<int, int> &startingPosition)
{
    string line;
    for (size_t rowIndex = 0; getline(cin, line); rowIndex++) {
        for (size_t colIndex = 0; colIndex < line.size(); colIndex++) {
            if (line[colIndex] == '^') {
                startingPosition = {rowIndex, colIndex};
            }
        }
        mappedArea.push_back(line);
    }
}

bool isOutOfBounds(const vector<string> &mappedArea, const pair<int, int> &position)
{
    return (position.first < 0 || position.first >= mappedArea.size())
           || (position.second < 0 || position.second >= mappedArea[0].size());
}

bool isObstacle(const vector<string> &mappedArea, const pair<int, int> &position)
{
    char cell = mappedArea[position.first][position.second];
    return cell == '#' || cell == 'O';
}

enum Direction { UP = 0, RIGHT, DOWN, LEFT };

bool takeStep(vector<string> &mappedArea,
              pair<int, int> &currentPosition,
              Direction &currentDirection,
              set<pair<int, int>> *distinctPositions = nullptr)
{
    const vector<pair<int, int>> directions = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
    pair<int, int> nextPosition = {currentPosition.first + directions[currentDirection].first,
                                   currentPosition.second + directions[currentDirection].second};

    if (isOutOfBounds(mappedArea, nextPosition)) {
        return false;
    }

    if (isObstacle(mappedArea, nextPosition)) {
        currentDirection = static_cast<Direction>((currentDirection + 1) % 4);
        return true;
    }

    if (distinctPositions && mappedArea[nextPosition.first][nextPosition.second] != 'X') {
        distinctPositions->insert({nextPosition.first, nextPosition.second});
    }

    mappedArea[currentPosition.first][currentPosition.second] = 'X';
    mappedArea[nextPosition.first][nextPosition.second] = '^';
    currentPosition = nextPosition;
    return true;
}

bool isLoopCausingObstruction(const pair<int, int> &obstructionPosition,
                              vector<string> mappedArea,
                              pair<int, int> currentPosition,
                              Direction currentDirection)
{
    if (obstructionPosition == currentPosition) {
        return false;
    }

    mappedArea[obstructionPosition.first][obstructionPosition.second] = 'O';

    set<tuple<int, int, Direction>> states;
    states.insert({currentPosition.first, currentPosition.second, currentDirection});

    while (takeStep(mappedArea, currentPosition, currentDirection)) {
        tuple<int, int, Direction> currentState = {currentPosition.first,
                                                   currentPosition.second,
                                                   currentDirection};

        if (states.count(currentState)) {
            return true;
        }

        states.insert(currentState);
    }

    return false;
}

int main()
{
    vector<string> mappedArea;
    pair<int, int> startingPosition;
    Direction startingDirection = UP;
    readInput(mappedArea, startingPosition);

    pair<int, int> currentPosition = startingPosition;
    Direction currentDirection = startingDirection;
    set<pair<int, int>> distinctPositions = {currentPosition};
    while (takeStep(mappedArea, currentPosition, currentDirection, &distinctPositions))
        ;
    int numberOfDistinctPositions = distinctPositions.size();

    int numberOfLoopCausingObstructions = 0;
    for (const pair<int, int> &obstructionPosition : distinctPositions) {
        numberOfLoopCausingObstructions += isLoopCausingObstruction(obstructionPosition,
                                                                    mappedArea,
                                                                    startingPosition,
                                                                    startingDirection);
    }

    // Part one
    cout << "Number of distinct positions visited before leaving: " << numberOfDistinctPositions
         << endl;

    // Part two
    cout << "Number of potential obstructions to create a loop: " << numberOfLoopCausingObstructions
         << endl;
}

File: 2024/go/day-06/main.go
package main

import (
	"fmt"
	"log"
	"os"
	"strings"
)

type position struct {
	row int
	col int
}

type state struct {
	row       int
	col       int
	direction int
}

const (
	up    = 0
	right = 1
	down  = 2
	left  = 3
)

func readInput() (area [][]rune, start state, err error) {
	content, err := os.ReadFile("input")
	if err != nil {
		return
	}

	lines := strings.Split(strings.TrimSpace(string(content)), "\n")
	area = make([][]rune, len(lines))

	for rowIndex, line := range lines {
		for colIndex, symbol := range line {
			if symbol == '^' {
				start = state{row: rowIndex, col: colIndex, direction: up}
			}
		}

		area[rowIndex] = []rune(line)
	}

	return
}

func stateToPos(s state) position {
	return position{s.row, s.col}
}

func isOutOfBounds(area [][]rune, pos position) bool {
	return (pos.row < 0 || pos.row >= len(area)) || (pos.col < 0 || pos.col >= len(area[0]))
}

func isObstacle(area [][]rune, pos position) bool {
	return area[pos.row][pos.col] == '#'
}

func takeStep(area [][]rune, currentState state) (bool, state) {
	directions := map[int]position{up: {-1, 0}, right: {0, 1}, down: {1, 0}, left: {0, -1}}

	newState := state{
		row:       currentState.row + directions[currentState.direction].row,
		col:       currentState.col + directions[currentState.direction].col,
		direction: currentState.direction,
	}

	if isOutOfBounds(area, stateToPos(newState)) {
		return false, newState
	}

	if isObstacle(area, stateToPos(newState)) {
		currentState.direction = (currentState.direction + 1) % 4
		return true, currentState
	}

	return true, newState
}

func findDistinctPositions(area [][]rune, currentState state) map[position]bool {
	distinctPositions := make(map[position]bool)

	for ok := true; ok; {
		distinctPositions[stateToPos(currentState)] = true

		ok, currentState = takeStep(area, currentState)
	}

	return distinctPositions
}

func isLoopCausingObstruction(area [][]rune, currentState state) bool {
	path := make(map[state]bool)

	for ok := true; ok; {
		if _, isLoop := path[currentState]; isLoop {
			return true
		}
		path[currentState] = true

		ok, currentState = takeStep(area, currentState)
	}

	return false
}

func findLoopCausingObstructions(area [][]rune, start state, distinctPositions map[position]bool) []position {
	var loopObstructions []position

	for obstaclePos := range distinctPositions {
		if obstaclePos == stateToPos(start) {
			continue
		}

		area[obstaclePos.row][obstaclePos.col] = '#'
		if isLoopCausingObstruction(area, start) {
			loopObstructions = append(loopObstructions, obstaclePos)
		}
		area[obstaclePos.row][obstaclePos.col] = '.'
	}

	return loopObstructions
}

func main() {
	area, start, err := readInput()
	if err != nil {
		log.Fatalln(err)
	}

	distinctPositions := findDistinctPositions(area, start)
	loopObstructions := findLoopCausingObstructions(area, start, distinctPositions)

	// Part one
	fmt.Println("Number of distinct positions:", len(distinctPositions))
	// Part two
	fmt.Println("Number of potential obstructions:", len(loopObstructions))
}
